airflow/api_fastapi/core_api/routes/ui/grid.py

```diff
@@ -0,0 +1,324 @@
+            # which is good enough in most cases.
+            if name in ("data_interval_start", "data_interval_end"):
+                expr = func.coalesce(expr, DagRun.logical_date)
+            return expr.desc()
+
+        ordering_expression = (_get_run_ordering_expr(name) for name in dag.timetable.run_ordering)
+        # create SortParam with ordering_expression and DagRun.id.desc()
+        return ordering_expression
+
+    dag_runs_select_filter, _ = paginated_select(
+        statement=base_query.order_by(*get_dag_run_sort_param(), DagRun.id.desc()),
+        filters=[
+            run_types,
+            run_states,
+        ],
+        order_by=None,
+        offset=offset,
+        limit=num_runs,
+    )
+
+    dag_runs = session.execute(dag_runs_select_filter)
+
+    # Check if there are any DAG Runs with given criteria to eliminate unnecessary queries/errors
+    if not dag_runs:
+        return GridResponse(dag_runs=[])
+
+    # Retrieve, sort and encode the Task Instances
+    tis_of_dag_runs, _ = paginated_select(
+        statement=select(
+            TaskInstance.run_id,
+            TaskInstance.task_id,
+            TaskInstance.try_number,
+            TaskInstance.state,
+            TaskInstance.start_date,
+            TaskInstance.end_date,
+            TaskInstance.queued_dttm.label("queued_dttm"),
+        )
+        .join(TaskInstance.task_instance_note, isouter=True)
+        .where(TaskInstance.dag_id == dag.dag_id),
+        filters=[],
+        order_by=SortParam(allowed_attrs=["task_id", "run_id"], model=TaskInstance).dynamic_depends(
+            "task_id"
+        )(),
+        offset=offset,
+        limit=None,
+    )
+
+    task_instances = session.execute(tis_of_dag_runs)
+
+    @cache
+    def get_task_group_children_getter() -> operator.methodcaller:
+        """Get the Task Group Children Getter for the DAG."""
+        sort_order = conf.get("webserver", "grid_view_sorting_order", fallback="topological")
+        if sort_order == "topological":
+            return operator.methodcaller("topological_sort")
+        if sort_order == "hierarchical_alphabetical":
+            return operator.methodcaller("hierarchical_alphabetical_sort")
+        raise AirflowConfigException(f"Unsupported grid_view_sorting_order: {sort_order}")
+
+    @cache
+    def get_task_group_map() -> dict[str, dict[str, Any]]:
+        """Get the Task Group Map for the DAG."""
+        task_nodes = {}
+
+        def _fill_task_group_map(
+            task_node: BaseOperator | MappedTaskGroup | TaskMap | None,
+            parent_node: BaseOperator | MappedTaskGroup | TaskMap | None,
+        ):
+            """Recursively fill the Task Group Map."""
+            if task_node is None:
+                return
+            if isinstance(task_node, MappedOperator):
+                task_nodes[task_node.node_id] = {
+                    "is_group": False,
```

Comment:
Actually, a task group can be mapped so I think we need an additional check here.

Reply:
Nice catch! I missed this thinking MappedTaskGroup is already a TaskGroup but you are right. I have included aggregated results under two instances by checking if MappedTaskGroup exists in each case. I compared the numbers in the old UI when hovering over the tasks and it seems the result is matching now.